Shopify/cli
# Send uninstall webhook when app id changes

### WHY are these changes introduced?

If a developer uses 2 different apps with the same store locally, the app thinks it is already installed to the store. That happens because the database entry isn't associated with the app itself, so the app decides it's installed even if it isn't actually installed in Shopify.

### WHAT is this pull request doing?

To help ensure that the app starts with a fresh state, the CLI will now send an `app/uninstalled` webhook to the app after it starts, so that it can clean up the state for the store, and reinstall an app if it changes.

This only happens IF AND ONLY IF:
- the user has just selected a new app, either by running `dev --reset` or passing in a different `--api-key` flag
- the app has a new `webhooks_path` value in their backend `shopify.web.toml` file

<details>
<summary>Example output</summary>

```
Using a different app than last time, sending uninstall webhook to app server


Shareable app URL

  <super secret URL>

2023-01-26 20:52:43 | backend  |
2023-01-26 20:52:43 | backend  | > dev
2023-01-26 20:52:43 | backend  | > cross-env NODE_ENV=development nodemon index.js --ignore ./frontend
2023-01-26 20:52:43 | backend  |
2023-01-26 20:52:43 | frontend |
2023-01-26 20:52:43 | frontend | > dev
2023-01-26 20:52:43 | frontend | > vite
2023-01-26 20:52:43 | frontend |
2023-01-26 20:52:43 | backend  | [nodemon] 2.0.20
2023-01-26 20:52:43 | backend  | [nodemon] to restart at any time, enter `rs`
2023-01-26 20:52:43 | backend  | [nodemon] watching path(s): *.*
2023-01-26 20:52:43 | backend  | [nodemon] watching extensions: js,mjs,json
2023-01-26 20:52:43 | backend  | [nodemon] starting `node index.js`
2023-01-26 20:52:43 | frontend |
2023-01-26 20:52:43 | frontend |   vite v2.9.15 dev server running at:
2023-01-26 20:52:43 | frontend |
2023-01-26 20:52:43 | frontend |   > Local:    http://localhost:61733/
2023-01-26 20:52:43 | frontend |
2023-01-26 20:52:43 | frontend |   ready in 152ms.
2023-01-26 20:52:43 | frontend |
2023-01-26 20:52:43 | backend  | [shopify-api/INFO] version 6.1.0, environment Node v18.7.0
2023-01-26 20:52:45 | webhooks | Sending APP_UNINSTALLED webhook to app server
2023-01-26 20:52:45 | backend  | [shopify-api/INFO] Processing webhook request
2023-01-26 20:52:45 | backend  | [shopify-app/INFO] Webhook processed, returned status code 200
2023-01-26 20:52:45 | webhooks | APP_UNINSTALLED webhook delivered
```

</details>

### How to test your changes?

You'll need 2 Shopify apps (A and B) for this. Steps:

- Create a new app
- Add `webhooks_path: /api/webhooks` to the backend TOML file
- Run `dev`, select app A
- Stop the app, run `dev --reset`, select app B
  - You should see the new behaviour here
- Stop the app again, run `dev --api-key=<API key for app A>`
  - Behaviour should also appear here
- Run `dev`
  - No webhook should be fired

### Post-release steps

Once this is released, we should add the default webhooks path to the app template TOML files.

### Measuring impact

How do we know this change was effective? Please choose one:


### Checklist

  - Is there a need to document this behaviour anywhere? It is a somewhat fringe case.
